File: py/src/tidyeavs/_aggregate.py
"""Quirk-aware rollups to state or county, with honest coverage.

Mirrors ``r/R/aggregate.R``. Named ``_aggregate`` rather than ``aggregate`` so the
module does not shadow the ``aggregate()`` function in the package namespace, the
same trap ``_dictionary`` avoids.
"""

from __future__ import annotations

import warnings
from typing import Any, Iterable

import numpy as np
import pandas as pd

from . import _metadata
from ._constants import TERRITORIES

# None means "skip the anomaly column", so a distinct sentinel is needed for
# "use the shipped table". Same pattern as _flags.py.
_DEFAULT = object()

_ID_COLUMNS = (
    "year",
    "survey",
    "fips_code",
    "jurisdiction_name",
    "state_abbr",
    "state_name",
    "state_fips",
    "county_fips",
)

# EAVS distinguishes why a value is absent, and the distinction decides whether a
# total is untrustworthy or perfectly complete. "Does not apply" and a valid skip
# mean the item did not pertain, so nothing is missing.
_NOT_APPLICABLE = ("does_not_apply", "valid_skip")


def aggregate(
    data: pd.DataFrame,
    status: pd.DataFrame | None = None,
    by: str = "state",
    concepts: Iterable[str] | None = None,
    jurisdictions: pd.DataFrame | None = None,
    anomalies: Any = _DEFAULT,
) -> pd.DataFrame:
    """Sum a harmonized panel to one row per group per concept.

    Reports how many jurisdictions stood behind each number rather than leaving
    you to assume all of them did. The output is long, one row per ``year``,
    group, and ``concept``, because a wide frame carrying a value plus five
    counts for each of ~40 concepts is unusable.

    **Coverage.** Pass ``status`` (a :func:`tidyeavs.missing_status` frame put
    through :func:`tidyeavs.harmonize`) and the counts distinguish *why* a
    jurisdiction is absent. That distinction matters: in 2024 a thousand
    jurisdictions had no provisional ballots at all, so counting them as gaps
    drops apparent coverage of ``prov_rejected`` from 94.5% to 79.9% when
    nothing is missing. Five buckets, which sum to ``n_total``:

    - ``n_reported`` gave a number, and are in ``value``.
    - ``n_missing`` are the genuine gaps (``not_available``,
      ``other_missing``): the ones that make a total untrustworthy.
    - ``n_blank`` left the cell empty, which the file does not explain. A blank
      could be a gap or a skip, so it is counted separately and the call is
      yours. Most blanks are 2016 Maine, whose municipalities leave the UOCAVA
      items to the statewide row.
    - ``n_not_applicable`` said the item did not pertain (``does_not_apply``,
      ``valid_skip``), so nothing is missing.
    - ``n_not_collected`` were never asked, because the concept was not on that
      year's questionnaire.

    ``coverage`` is ``n_reported / (n_reported + n_missing + n_blank)``: the
    share of jurisdictions that could have answered and did. The last two
    buckets stay out of the denominator, since nothing is missing there.
    Without ``status`` the reason is unknowable after recoding, so every
    absence lands in ``n_missing`` and ``coverage`` is a lower bound;
    ``coverage_exact`` says which case you have.

    ``coverage_reg`` weights the same ratio by ``reg_eligible_total``,
    answering how much of the electorate the reporters hold rather than how
    many of them there were. Since ``value`` is a sum, that is usually the
    better guide to how much of it you have: in 2024 ``ballots_cured`` comes
    from 28.5% of jurisdictions but 65.9% of the registered electorate.
    Jurisdictions with no usable weight leave both sides of the ratio, and the
    column is missing if the panel carries no ``reg_eligible_total``.

    Every row also carries ``entity_type`` (``"state"``, ``"territory"``, or
    ``"district"``), so a 50-state analysis is one filter.

    **Anomalous state-years.** ``known_anomaly`` is ``True`` where this year,
    state, and concept are recorded in :func:`tidyeavs.known_anomalies` and at
    least one jurisdiction in the group reported a value, so the total rests on
    the anomalous convention. Coverage cannot tell you this: Iowa's 2018 polling
    places are reported by all 99 counties, so every coverage column reads as
    complete, and the state total is still not comparable to its other cycles.
    The value is summed as reported either way; :func:`tidyeavs.flags` gives the
    reason and the evidence. Pass ``anomalies=None`` to skip the column's
    lookup.

    **Nothing is corrected.** Values are summed as reported, and ``NA`` is
    skipped rather than imputed, so a "does not apply" never becomes a zero.
    Maine's statewide UOCAVA row and the territories are included, because both
    carry real reported data; filter them afterwards if your analysis wants
    that.

    ``by="county"`` drops rows whose published code embeds no county, and says
    how many: Wisconsin's serials, Maine's statewide row, Alaska, the
    territories. Wisconsin stays out even though
    :func:`tidyeavs.jurisdictions` parses its county names, because the ~57
    municipalities that straddle county lines include the City of Milwaukee.
    """
    if by not in ("state", "county"):
        raise ValueError(f"by must be 'state' or 'county'; got {by!r}")
    if "year" not in data.columns:
        raise ValueError("data needs a 'year' column; use tidyeavs.load().")
    if anomalies is _DEFAULT:
        anomalies = _metadata.known_anomalies()
    if status is not None:
        if len(status) != len(data):
            raise ValueError(
                f"status must line up row-for-row with data: data has {len(data)} rows, "
                f"status has {len(status)}. Harmonize both from the same read()."
            )
        # Matching row counts are not alignment: binding per-year status frames
        # in a different year order than the panel passes the count check and
        # then reports another year's coverage. Both frames carry the keys.
        for key in ("year", "fips_code"):
            if key in data.columns and key in status.columns:
                left = data[key].astype("string").to_numpy()
                right = status[key].astype("string").to_numpy()
                if not (left == right).all():
                    raise ValueError(
                        f"status is not aligned with data: their {key!r} columns "
                        "differ row-for-row. Bind the per-year status frames in "
                        "the same year order as load()."
                    )

    items = _item_columns(data, concepts)
    keys, keep = _group_keys(data, by, jurisdictions)
    # The coverage_reg weight, taken from the panel itself whether or not
    # reg_eligible_total is among the concepts being rolled up.
    weight = data["reg_eligible_total"] if "reg_eligible_total" in data.columns else None

    frames = [
        _aggregate_one(
            concept,
            data[concept],
            status[concept] if (status is not None and concept in status.columns) else None,
            keys,
            keep,
            weight,
        )
        for concept in items
    ]
    out = pd.concat(frames, ignore_index=True)
    out["entity_type"] = _entity_type(out["state_abbr"])
    # A total can rest on an anomalous reporting convention while every coverage
    # column reads as complete, so say so here rather than leaving it to a
    # separate flags() call the user has to know to make.
    out["known_anomaly"] = _anomaly_match(
        out["year"], out["state_abbr"], out["concept"], anomalies
    ) & (out["n_reported"] > 0).to_numpy()

    lead = ["year"] + [c for c in keys.columns if c != "year"]
    out = out[
        lead
        + [
            "entity_type",
            "concept",
            "value",
            "n_reported",
            "n_missing",
            "n_blank",
            "n_not_applicable",
            "n_not_collected",
            "n_total",
            "coverage",
            "coverage_reg",
            "coverage_exact",
            "known_anomaly",
        ]
    ]
    return out.sort_values(["year", "concept"] + [c for c in lead if c != "year"],
                          kind="mergesort").reset_index(drop=True)


def _item_columns(data: pd.DataFrame, concepts: Iterable[str] | None) -> list[str]:
    candidates = [c for c in data.columns if c not in _ID_COLUMNS]
    items = [c for c in candidates if pd.api.types.is_numeric_dtype(data[c])]

    if concepts is not None:
        wanted = list(concepts)
        unknown = [c for c in wanted if c not in items]
        if unknown:
            raise ValueError(
                f"Not numeric concept column(s) in data: {unknown}. Available: {items}."
            )
        items = wanted
    if not items:
        raise ValueError(
            "No numeric concept columns found in data. Did you forget "
            "recode_missing() before harmonizing?"
        )
    return items


def _entity_type(state_abbr: pd.Series) -> pd.Series:
    """Classify a state_abbr column as state, territory, or district (DC)."""
    out = pd.Series("state", index=state_abbr.index, dtype="object")
    out[state_abbr.isin(TERRITORIES)] = "territory"
    out[state_abbr == "DC"] = "district"
    return out


def _anomaly_match(year, state_abbr, concept, anomalies) -> np.ndarray:
    """Is this year/state/concept one of the verified reporting anomalies?

    Keyed the same three ways ``_flag_known_anomalies`` matches, since every
    anomaly recorded so far is a statewide convention for one concept in one
    cycle. ``concept`` may be a single name, recycled. Mirrors
    ``anomaly_match()`` in ``utils.R``.

    This is what lets :func:`rate` and :func:`aggregate` say that a number rests
    on an anomalous state-year, which no arithmetic check can tell you: Oregon's
    2018 mail rejection rate comes back with every county reporting and is still
    unusable. Reporting it is as far as this goes—the value is returned as
    summed, and what to do about it is the user's call.
    """
    years = list(pd.to_numeric(pd.Series(list(year)), errors="coerce").astype("Int64"))
    states = list(state_abbr)
    n = len(years)
    concepts = [concept] * n if isinstance(concept, str) else list(concept)

    if anomalies is None or len(anomalies) == 0:
        return np.zeros(n, dtype=bool)

    known = {
        f"{a_year} {a_state} {a_concept}"
        for a_year, a_state, a_concept in zip(
            pd.to_numeric(anomalies["year"], errors="coerce").astype("Int64"),
            anomalies["state_abbr"],
            anomalies["concept"],
        )
    }
    return np.array(
        [f"{y} {s} {c}" in known for y, s, c in zip(years, states, concepts)],
        dtype=bool,
    )


def _group_keys(
    data: pd.DataFrame, by: str, jurisdictions: pd.DataFrame | None
) -> tuple[pd.DataFrame, pd.Series]:
    if "state_abbr" not in data.columns:
        raise ValueError("data needs a 'state_abbr' column to group by.")

    keys = pd.DataFrame(
        {"year": data["year"].to_numpy(), "state_abbr": data["state_abbr"].to_numpy()},
        index=data.index,
    )
    if by == "state":
        return keys, pd.Series(True, index=data.index)

    if "fips_code" not in data.columns:
        raise ValueError("A county rollup needs a 'fips_code' column in data.")
    jur = _metadata.jurisdictions() if jurisdictions is None else jurisdictions

    lookup = jur.set_index(
        [jur["year"].astype("Int64"), jur["fips_code"].astype("string")]
    )["county_fips"]
    # A few Wisconsin serials are shared by two published rows in one year, so
    # this index is not unique and reindex would refuse it. Both rows of a pair
    # carry the same county, so taking the first matches what R's match() does.
    lookup = lookup[~lookup.index.duplicated()]
    probe = pd.MultiIndex.from_arrays(
        [data["year"].astype("Int64"), data["fips_code"].astype("string")]
    )
    keys["county_fips"] = lookup.reindex(probe).to_numpy()

    keep = keys["county_fips"].notna()
    dropped = int((~keep).sum())
    if dropped:
        warnings.warn(
            f"Dropping {dropped} row(s) with no county in the published code: "
            "Wisconsin's non-geographic serials, Maine's statewide row, Alaska, "
            "and the territories have none.",
            stacklevel=3,
        )
    return keys, keep


def _aggregate_one(
    concept: str,
    value: pd.Series,
    reason: pd.Series | None,
    keys: pd.DataFrame,
    keep: pd.Series,
    weight: pd.Series | None = None,
) -> pd.DataFrame:
    exact = reason is not None
    work = keys[keep].copy()
    values = pd.to_numeric(value[keep], errors="coerce")

    if exact:
        labels = reason[keep].astype("string")
        reported = labels == "reported"
        not_applicable = labels.isin(_NOT_APPLICABLE)
        # A blank is ambiguous—gap or skip, the file does not say—so it is
        # counted apart rather than folded into either. See the docstring.
        blank = labels == "blank"
        # No reason at all means the concept was not on that year's
        # questionnaire (a multi-year status bind fills those rows with NA),
        # which gets its own bucket so the five sum to n_total.
        not_collected = labels.isna()
    else:
        # After recoding the reason is gone, so every absence counts as a gap.
        reported = values.notna()
        not_applicable = pd.Series(False, index=values.index)
        blank = pd.Series(False, index=values.index)
        not_collected = pd.Series(False, index=values.index)

    work["_value"] = values.astype("float64")
    work["_reported"] = reported.fillna(False).astype(int).to_numpy()
    work["_na"] = not_applicable.fillna(False).astype(int).to_numpy()
    work["_blank"] = blank.fillna(False).astype(int).to_numpy()
    work["_nc"] = not_collected.astype(int).to_numpy()
    work["_gap"] = (
        1 - work["_reported"] - work["_na"] - work["_blank"] - work["_nc"]
    ).clip(lower=0)

    # Registration-weighted coverage: how much of the electorate the reporting
    # jurisdictions hold. Rows with no usable weight contribute to neither side.
    if weight is not None:
        w = pd.to_numeric(weight[keep], errors="coerce").fillna(0).astype("float64")
        work["_w_reported"] = w.to_numpy() * work["_reported"]
        work["_w_answerable"] = w.to_numpy() * (
            work["_reported"] + work["_gap"] + work["_blank"]
        )
    else:
        work["_w_reported"] = 0.0
        work["_w_answerable"] = 0.0

    group_cols = list(keys.columns)
    grouped = work.groupby(group_cols, dropna=False, sort=True)
    out = grouped.agg(
        value=("_value", "sum"),
        n_reported=("_reported", "sum"),
        n_missing=("_gap", "sum"),
        n_blank=("_blank", "sum"),
        n_not_applicable=("_na", "sum"),
        n_not_collected=("_nc", "sum"),
        n_total=("_value", "size"),
        _w_reported=("_w_reported", "sum"),
        _w_answerable=("_w_answerable", "sum"),
    ).reset_index()

    out["concept"] = concept
    answerable = out["n_reported"] + out["n_missing"] + out["n_blank"]
    out["coverage"] = (out["n_reported"] / answerable).where(answerable > 0)
    if weight is not None:
        out["coverage_reg"] = (out["_w_reported"] / out["_w_answerable"]).where(
            out["_w_answerable"] > 0
        )
    else:
        out["coverage_reg"] = float("nan")
    out = out.drop(columns=["_w_reported", "_w_answerable"])
    out["coverage_exact"] = exact
    return out

File: r/R/aggregate.R
#' Roll EAVS jurisdictions up to states, with honest coverage
#'
#' Sums a harmonized panel (see [eavs_load()]) to one row per group per concept,
#' and reports how many jurisdictions stood behind each number rather than
#' leaving you to assume all of them did.
#'
#' The output is long, one row per `year`, group, and `concept`, because a wide
#' frame carrying a value plus five counts for each of ~40 concepts is unusable.
#' Pivot it yourself if you want concepts as columns.
#'
#' @section Coverage:
#' Pass `status` (a [eavs_missing_status()] frame put through
#' [eavs_harmonize()]) and the counts distinguish *why* a jurisdiction is
#' absent. That distinction matters: EAVS separates "does not apply" from "data
#' not available", and treating them alike misreports coverage in both
#' directions. In 2024 a thousand jurisdictions had no provisional ballots at
#' all, so counting them as gaps drops apparent coverage of `prov_rejected`
#' from 94.5% to 79.9% when nothing is missing.
#'
#' Five buckets, which sum to `n_total`:
#'
#' - `n_reported` gave a number, and are in `value`.
#' - `n_missing` are the genuine gaps (`not_available`, `other_missing`): the
#'   ones that make a total untrustworthy.
#' - `n_blank` left the cell empty, which the file does not explain. A blank
#'   could be a gap or a skip, so it is counted separately and the call is
#'   yours. Most blanks are 2016 Maine, whose municipalities leave the UOCAVA
#'   items to the statewide row.
#' - `n_not_applicable` said the item did not pertain (`does_not_apply`,
#'   `valid_skip`), so nothing is missing.
#' - `n_not_collected` were never asked, because the concept was not on that
#'   year's questionnaire.
#'
#' `coverage` is `n_reported / (n_reported + n_missing + n_blank)`: the share of
#' jurisdictions that could have answered and did. The last two buckets stay out
#' of the denominator, since nothing is missing there. Without `status` the
#' reason is unknowable after recoding, so every absence lands in `n_missing`
#' and `coverage` is a lower bound; `coverage_exact` says which case you have.
#'
#' `coverage_reg` weights the same ratio by `reg_eligible_total`, answering how
#' much of the electorate the reporters hold rather than how many of them there
#' were. Since `value` is a sum, that is usually the better guide to how much of
#' it you have: in 2024 `ballots_cured` comes from 28.5% of jurisdictions but
#' 65.9% of the registered electorate. Jurisdictions with no usable weight leave
#' both sides of the ratio, and the column is `NA` if the panel carries no
#' `reg_eligible_total`.
#'
#' @section Anomalous state-years:
#' `known_anomaly` is `TRUE` where this year, state, and concept are recorded in
#' [eavs_known_anomalies] and at least one jurisdiction in the group reported a
#' value, so the total rests on the anomalous convention. Coverage cannot tell
#' you this: Iowa's 2018 polling places are reported by all 99 counties, so every
#' coverage column reads as complete, and the state total is still not comparable
#' to its other cycles. The value is summed as reported either way, and what to
#' do about it is yours; [eavs_flags()] gives the reason and the evidence.
#'
#' @section What is not corrected:
#' Nothing. Values are summed as reported, and `NA` is skipped rather than
#' imputed, so a "does not apply" never becomes a zero. Maine's statewide UOCAVA
#' row and the territories are included, because both carry real reported data.
#' Filter them afterwards if your analysis wants that; `entity_type` and
#' [eavs_jurisdictions] make it a one-liner.
#'
#' @param data A harmonized panel from [eavs_load()], or one year from
#'   [eavs_harmonize()].
#' @param status Optional. The matching [eavs_missing_status()] output, harmonized
#'   the same way, which makes the coverage counts reason-aware. It must line up
#'   with `data` row for row; `year` and `fips_code` are checked, so a frame
#'   bound in a different year order is an error rather than wrong coverage.
#' @param by `"state"` (default) or `"county"`. County rollups drop rows whose
#'   published code embeds no county, and say how many: Wisconsin's serials,
#'   Maine's statewide row, Alaska, the territories. Wisconsin stays out even
#'   though [eavs_jurisdictions] parses its county names, because the ~57
#'   municipalities that straddle county lines include the City of Milwaukee.
#' @param concepts Optional character vector limiting which concepts to roll up.
#' @param jurisdictions The jurisdiction table to join for county codes. Defaults
#'   to the bundled [eavs_jurisdictions].
#' @param anomalies The verified reporting anomalies to mark. Defaults to the
#'   bundled [eavs_known_anomalies]; pass your own or set to `NULL` to skip them.
#'
#' @return A tibble with one row per group per concept per year: the grouping
#'   columns, `entity_type` (`"state"`, `"territory"`, or `"district"`, so a
#'   50-state analysis is one filter), `concept`, `value`, the five bucket
#'   counts, `n_total`, `coverage`, `coverage_reg`, `coverage_exact`, and
#'   `known_anomaly`.
#' @seealso [eavs_rate()] for a rate rather than a total, [eavs_flags()] for
#'   consistency checks, [eavs_missing_status()] for per-jurisdiction reasons.
#' @export
#'
#' @examples
#' \dontrun{
#' panel <- eavs_load(2024)
#' eavs_aggregate(panel)
#'
#' # Reason-aware coverage needs the status frame alongside:
#' raw <- eavs_read(2024)
#' st <- eavs_harmonize(eavs_missing_status(raw), year = 2024)
#' eavs_aggregate(panel, status = st)
#' }
eavs_aggregate <- function(data, status = NULL, by = "state", concepts = NULL,
                           jurisdictions = NULL, anomalies = NULL) {
  by <- match.arg(by, c("state", "county"))
  if (missing(anomalies)) {
    anomalies <- eavs_known_anomalies
  }

  if (!"year" %in% names(data)) {
    cli::cli_abort("{.arg data} needs a {.field year} column; use {.fn eavs_load}.")
  }
  if (!is.null(status)) {
    if (nrow(status) != nrow(data)) {
      cli::cli_abort(c(
        "{.arg status} must line up row-for-row with {.arg data}.",
        x = "{.arg data} has {nrow(data)} rows, {.arg status} has {nrow(status)}.",
        i = "Harmonize both from the same {.fn eavs_read} call."
      ))
    }
    # Matching row counts are not alignment: binding per-year status frames in a
    # different year order than the panel passes the count check and then
    # reports another year's coverage. Both frames carry the keys to prove it.
    for (key in intersect(c("year", "fips_code"), intersect(names(data), names(status)))) {
      if (!identical(as.character(data[[key]]), as.character(status[[key]]))) {
        cli::cli_abort(c(
          "{.arg status} is not aligned with {.arg data}.",
          x = "Their {.field {key}} columns differ row-for-row.",
          i = "Bind the per-year status frames in the same year order as {.fn eavs_load}."
        ))
      }
    }
  }

  item_cols <- aggregate_items(data, concepts)
  keys <- aggregate_keys(data, by, jurisdictions)
  # The coverage_reg weight, taken from the panel itself whether or not
  # reg_eligible_total is among the concepts being rolled up.
  weight <- if ("reg_eligible_total" %in% names(data)) {
    data$reg_eligible_total
  } else {
    NULL
  }

  out <- vector("list", length(item_cols))
  for (i in seq_along(item_cols)) {
    out[[i]] <- aggregate_one(
      concept = item_cols[i],
      value = data[[item_cols[i]]],
      reason = if (is.null(status) || !item_cols[i] %in% names(status)) {
        NULL
      } else {
        as.character(status[[item_cols[i]]])
      },
      keys = keys,
      weight = weight
    )
  }
  res <- dplyr::bind_rows(out)

  res$entity_type <- entity_type_of(res$state_abbr)
  # A total can rest on an anomalous reporting convention while every coverage
  # column reads as complete, so say so here rather than leaving it to a
  # separate eavs_flags() call the user has to know to make.
  res$known_anomaly <- anomaly_match(res$year, res$state_abbr, res$concept,
                                     anomalies) & res$n_reported > 0

  key_names <- setdiff(names(keys), ".row")
  res <- res[order(res$year, res$concept), c("year", setdiff(key_names, "year"),
                                             "entity_type", "concept", "value",
                                             "n_reported", "n_missing",
                                             "n_blank", "n_not_applicable",
                                             "n_not_collected",
                                             "n_total", "coverage",
                                             "coverage_reg",
                                             "coverage_exact",
                                             "known_anomaly")]
  tibble::as_tibble(res)
}

# Which columns are concepts to sum, as opposed to identifiers?
aggregate_items <- function(data, concepts = NULL) {
  ids <- c("year", "survey", "fips_code", "jurisdiction_name", "state_abbr",
           "state_name", "state_fips", "county_fips")
  candidates <- setdiff(names(data), ids)
  numeric_ok <- vapply(candidates, function(nm) is.numeric(data[[nm]]), logical(1))
  items <- candidates[numeric_ok]

  if (!is.null(concepts)) {
    unknown <- setdiff(concepts, items)
    if (length(unknown) > 0) {
      cli::cli_abort(c(
        "Not numeric concept column{?s} in {.arg data}: {.val {unknown}}.",
        i = "Available: {.val {items}}."
      ))
    }
    items <- concepts
  }
  if (length(items) == 0) {
    cli::cli_abort(c(
      "No numeric concept columns found in {.arg data}.",
      i = "Did you forget {.fn eavs_recode_missing} before harmonizing?"
    ))
  }
  items
}

# The grouping columns, joining the jurisdiction table when rolling up to county.
aggregate_keys <- function(data, by, jurisdictions = NULL) {
  if (!"state_abbr" %in% names(data)) {
    cli::cli_abort("{.arg data} needs a {.field state_abbr} column to group by.")
  }
  keys <- tibble::tibble(year = data$year, state_abbr = data$state_abbr)
  if (by == "state") {
    return(keys)
  }

  jur <- if (is.null(jurisdictions)) eavs_jurisdictions else jurisdictions
  if (!"fips_code" %in% names(data)) {
    cli::cli_abort("A county rollup needs a {.field fips_code} column in {.arg data}.")
  }
  idx <- match(
    paste(data$year, data$fips_code),
    paste(jur$year, jur$fips_code)
  )
  keys$county_fips <- jur$county_fips[idx]

  unmatched <- sum(is.na(idx))
  if (unmatched > 0) {
    cli::cli_warn(
      "{unmatched} row{?s} did not match {.arg jurisdictions} on year + fips_code."
    )
  }
  dropped <- sum(is.na(keys$county_fips))
  if (dropped > 0) {
    cli::cli_inform(c(
      "!" = "Dropping {dropped} row{?s} with no county in the published code.",
      i = paste("Wisconsin's non-geographic serials, Maine's statewide row,",
                "Alaska, and the territories have none.")
    ))
  }
  keys
}

# Sum one concept within groups, counting why jurisdictions are absent.
aggregate_one <- function(concept, value, reason, keys, weight = NULL) {
  keep <- if ("county_fips" %in% names(keys)) !is.na(keys$county_fips) else TRUE
  keys <- keys[keep, , drop = FALSE]
  value <- value[keep]
  if (!is.null(reason)) {
    reason <- reason[keep]
  }
  if (!is.null(weight)) {
    weight <- weight[keep]
  }

  group <- interaction(keys, drop = TRUE, lex.order = TRUE)
  exact <- !is.null(reason)
  none <- rep(FALSE, length(value))

  # Without a status frame the reason is gone, so every absence is a gap.
  # With one, an NA reason means the concept was not on that year's
  # questionnaire (a multi-year status bind fills those rows with NA), which
  # gets its own bucket so the five sum to n_total.
  is_not_collected <- if (exact) is.na(reason) else none
  is_reported <- if (exact) !is_not_collected & reason == "reported" else !is.na(value)
  is_na_ok <- if (exact) reason %in% c("does_not_apply", "valid_skip") else none
  # A blank is ambiguous—gap or skip, the file does not say—so it is counted
  # apart rather than folded into either. See the Coverage section.
  is_blank <- if (exact) !is_not_collected & reason == "blank" else none
  is_gap <- !is_reported & !is_na_ok & !is_blank & !is_not_collected

  agg <- function(x) as.integer(tapply(x, group, sum, na.rm = TRUE))
  totals <- as.numeric(tapply(value, group, sum, na.rm = TRUE))

  first_rows <- match(levels(group), as.character(group))
  res <- keys[first_rows, , drop = FALSE]
  res$concept <- concept
  res$value <- totals
  res$n_reported <- agg(is_reported)
  res$n_missing <- agg(is_gap)
  res$n_blank <- agg(is_blank)
  res$n_not_applicable <- agg(is_na_ok)
  res$n_not_collected <- agg(is_not_collected)
  res$n_total <- as.integer(table(group))
  answerable <- res$n_reported + res$n_missing + res$n_blank
  res$coverage <- ifelse(answerable > 0, res$n_reported / answerable, NA_real_)

  # Registration-weighted coverage: how much of the electorate the reporting
  # jurisdictions hold. Rows with no usable weight contribute to neither side.
  if (is.null(weight)) {
    res$coverage_reg <- NA_real_
  } else {
    w <- ifelse(is.na(weight), 0, as.numeric(weight))
    wagg <- function(flag) as.numeric(tapply(ifelse(flag, w, 0), group, sum, na.rm = TRUE))
    w_reported <- wagg(is_reported)
    w_answerable <- w_reported + wagg(is_gap) + wagg(is_blank)
    res$coverage_reg <- ifelse(w_answerable > 0, w_reported / w_answerable, NA_real_)
  }
  res$coverage_exact <- exact
  res
}
